### PR TITLE
Fix: refactor constant contact provider

### LINKED ISSUE #503: Not able to create newsletter draft with Constant Contact provider
While attempting to start a new newsletter with Constant Contact as my provider, I get the following fatal error:

```
Fatal error: Uncaught TypeError: Argument 3 passed to GuzzleHttp\Client::request() must be of the type array, string given, called in newspack-plugin/vendor/guzzlehttp/guzzle/src/Client.php on line 92 and defined in newspack-plugin/vendor/guzzlehttp/guzzle/src/Client.php on line 184

Call Stack
--
1 | 0.0001 | 375496 | {main}( ) | .../post-new.php:0
2 | 0.2890 | 12181408 | get_default_post_to_edit( ) | .../post-new.php:66
3 | 0.2890 | 12181784 | wp_insert_post( ) | .../post.php:690
4 | 0.2947 | 12249032 | do_action( ) | .../post.php:4373
5 | 0.2947 | 12249408 | WP_Hook->do_action( ) | .../plugin.php:484
6 | 0.2947 | 12249408 | WP_Hook->apply_filters( ) | .../class-wp-hook.php:316
7 | 0.2963 | 12244056 | Newspack_Newsletters_Constant_Contact->save( ) | .../class-wp-hook.php:292
8 | 0.2963 | 12244056 | Newspack_Newsletters_Constant_Contact->sync( ) | .../class-newspack-newsletters-constant-contact.php:374
9 | 0.2980 | 12402720 | Ctct\Services\AccountService->getAccountInfo( ) | .../class-newspack-newsletters-constant-contact.php:314
10 | 0.2981 | 12402816 | Ctct\Services\AccountService->createBaseRequest( ) | .../AccountService.php:94
11 | 0.2981 | 12402816 | GuzzleHttp\Client->createRequest( ) | .../BaseService.php:65
12 | 0.2981 | 12403192 | GuzzleHttp\Client->__call( ) | .../BaseService.php:65
13 | 0.2981 | 12403192 | GuzzleHttp\Client->request( ) | .../Client.php:92
```

### LINKED ISSUE #511: Unable to send test email using Constant Contact
While attempting to send a test email with Constant Contact as my provider I get the following error message:

```
Server error response [url] https://api.constantcontact.com/v2/emailmarketing/campaigns/xxx/tests?api_key=xxx [status code] 500 [reason phrase] Internal Server Error
```

### LINKED ISSUE #512: Inconsistent state on "Newsletter" section while editing with Constant Contact
While editing a newsletter using Constant Contact as my provider, I get an inconsistent "sent" state on the "Newsletter" section of the editor.

## Steps to reproduce

1. Create a newsletter with Constant Contact as your provider
2. Setup the Newsletter section of the editor with appropriate values to send
3. Click on Send and wait for the prompt to confirm submission
4. Cancel submission and observe that the section has been updated with a "sent" notice (first inconsistency)
5. Click on Send again and, this time, confirm submission
6. Update or wait for an autosave and observe that the "sent" notice disappears and the section allows you to reconfigure your submission settings (second inconsistency)

### PR DESCRIPTION
**Pull request by miguelpeixe**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Implements our own [simplified SDK](https://github.com/Automattic/newspack-newsletters/blob/fix/refactor-constant-contact/includes/service-providers/constant_contact/class-newspack-newsletters-constant-contact-sdk.php) to handle the [maintained v3 API](https://v3.developer.constantcontact.com/api_guide/index.html) with OAuth integration.

Instead of storing a generated access token, the API credentials now store API's `key` and `secret`, used to manage access and refresh tokens. After the first configuration with the app's credentials, the editor sidebar is replaced with an Authorization button, which will prompt the OAuth flow, update the tokens seamlessly for the user and synchronize the post with the provider.

I struggled a little bit to implement the ability to synchronize the newsletter with the provider **after** the draft was created (either by the post being created with a different provider or before the tokens were available). I wasn't able to identify the root of this sync issue, but I identified that it was caused by the `cc_campaign_id` meta being available on the rest API. At some save cycle, the meta was being deleted and the post was desynced from the provider. Removing the meta from the API (see 2b1061b47fc63c37118105393709a26f7475142e) solved the issue. I figured it's not a problem, since we don't (and probably shouldn't) manage this data on the frontend.

#### Composer Autoloader update

To ensure that our Guzzle dependency does not conflict with an older version loaded from a different plugin, I'm using [this autoloader by Jetpack](https://github.com/Automattic/jetpack/tree/master/projects/packages/autoloader). According to its readme:

> This is a custom autoloader generator that uses a classmap to always load the latest version of a class.

Might be the case for further testing, but seems to work great here.

Closes #503.
Closes #511.
Closes #512.

### How to test the changes in this Pull Request:

1. Register a [Constant Contact Application](https://app.constantcontact.com/pages/dma/portal/)
2. Copy the API key and the generated secret
3. Update your Redirect URI setting to include: `https://newspack.local/wp-admin/index.php?cc_oauth2callback=*`
3. Make sure you have a Constant Contact account with updated information (or the API might fail sending)
4. Check out this branch and run `composer install && npm run build`
5. Go to Newsletters settings and set the provider to Constant Contact with your API credentials
6. Create a new newsletter and select a layout
7. Observe the "Authorize" message and button
8. Click on Authorize, login to your account if necessary, and Allow the app
9. Observe that the sidebar has been updated to synchronize with the provider
10. When it finishes retrieving the data, test by updating the content, sending test emails, and sending the newsletter.

In case your Constant Contact account is under review or suspended, you'll receive the following error notice while trying to send a test email or publishing:

Captura de Tela 2021-07-27 às 11 33 31

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Review by dkoo — CHANGES_REQUESTED** [review]
This is impressive work on such a short timeline! After reauthenticating, the integration is working as described. I like how you handle the OAuth flow right in the Newsletter editor sidebar—makes it feel as frictionless as possible, especially since you should only need to do this once. 

I tested: 

- Creation of new drafts
- Sending test emails
- Sending live emails
- Scheduling live emails

All worked in the end, although as discussed when I tried to send my first draft, I ran into trouble due to missing physical address info on my CC account. After updating my account info, the draft remained unsendable because the error info is saved on the draft campaign even after the account is updated. This should theoretically be fixed by 00874de, which sends physical address info with every send request, but I'm unable to verify at the moment because the original draft is no longer syncing to CC and my CC account is no longer in a state which would trigger the error. Still, this is not a blocker since the issue can be bypassed by copying the newsletter content into a brand new draft, which can be sent without issue.

I ran into a couple of PHP warnings, details below. After fixing those, this can be merged.

File: includes/service-providers/constant_contact/class-newspack-newsletters-constant-contact.php [hunk 1]
```diff
@@ -320,38 +521,40 @@ public function sync( $post ) {
...
+					function ( $email ) {
+						return 'CONFIRMED' === $email->confirm_status;
+					}
+				);
 
 				if ( empty( $verified_email_addresses ) ) {
 					throw new Exception( __( 'There are no verified email addresses in the Constant Contact account.', 'newspack-newsletters' ) );
 				}
 
 				$initial_email_address = $verified_email_addresses[0]->email_address;
```
[the remark this answers is not in the document]
Comment (miguelpeixe) [reply includes/service-providers/constant_contact/class-newspack-newsletters-constant-contact.php:540]: This is most likely due to `array_filter` not reindexing the array values. I'll call `array_values` on the filtered result so we can ensure that `[0]` exists.

File: includes/service-providers/constant_contact/class-newspack-newsletters-constant-contact.php [hunk 2]
```diff
@@ -371,6 +574,9 @@ public function save( $post_id, $post, $update ) {
 		if ( 'trash' === $status ) {
 			return;
 		}
+		if ( defined( 'DOING_AUTOSAVE' ) && DOING_AUTOSAVE ) {
```
[the remark this answers is not in the document]
Comment (miguelpeixe) [reply includes/service-providers/constant_contact/class-newspack-newsletters-constant-contact.php:581]: Yes. We should also bail sync on autosaves for other ESPs, there's no need for these extra API calls if we sync before every send/test.

[the remark this answers is not in the document]
Comment (miguelpeixe) [reply includes/service-providers/constant_contact/class-newspack-newsletters-constant-contact.php:581]: Just filed #524 for this.

File: includes/service-providers/constant_contact/class-newspack-newsletters-constant-contact.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (miguelpeixe) [reply includes/service-providers/constant_contact/class-newspack-newsletters-constant-contact.php:542]: Please verify if 17deab4440b618ccc71c1a1d76d0c78a4fb8a064 fixes the issue.

File: includes/service-providers/constant_contact/class-newspack-newsletters-constant-contact.php [hunk 3]
```diff
@@ -371,6 +574,9 @@ public function save( $post_id, $post, $update ) {
...
+			return;
+		}
 		$this->sync( $post );
```
[the remark this answers is not in the document]
Comment (miguelpeixe) [reply includes/service-providers/constant_contact/class-newspack-newsletters-constant-contact.php:586]: This most likely is also fixed by 17deab4440b618ccc71c1a1d76d0c78a4fb8a064. The non-object is the missing `[0]` index from the filtered array.

File: includes/service-providers/constant_contact/class-newspack-newsletters-constant-contact.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (dkoo) [reply includes/service-providers/constant_contact/class-newspack-newsletters-constant-contact.php:583]: Thanks for the new issue!

File: includes/service-providers/constant_contact/class-newspack-newsletters-constant-contact.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (dkoo) [reply includes/service-providers/constant_contact/class-newspack-newsletters-constant-contact.php:586]: Yes, can confirm that this is now fixed.

File: includes/service-providers/constant_contact/class-newspack-newsletters-constant-contact.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (dkoo) [reply includes/service-providers/constant_contact/class-newspack-newsletters-constant-contact.php:542]: Yes, can confirm that this is now fixed.

**Review by dkoo — APPROVED** [review]
Looking good! 🙌

**Comment by kbrown9** [comment]
Hi!

> Composer Autoloader update
To ensure that our Guzzle dependency does not conflict with an older version loaded from a different plugin, I'm using this autoloader by Jetpack. According to its readme:

> >This is a custom autoloader generator that uses a classmap to always load the latest version of a class.

There are some limitations to the Jetpack Autoloader that you should be aware of.

The Jetpack Autoloader can only control which version is loaded when every plugin with that dependency uses the Jetpack Autoloader. If any plugin on the site uses the dependency but does not use the Jetpack Autoloader (for example, if it uses the Composer autoloader instead), the JP Autoloader can't fully control which version gets loaded. So it's possible that you will still run into situations where an older version of Guzzle is loaded first.

I'm planning to update the Jetpack Autoloader's README to make sure this limitation is clear.

Another approach to consider is adding a custom namespace to dependencies. For example, Yoast adds a namespace to some of its dependencies, including Guzzle: https://plugins.trac.wordpress.org/browser/wordpress-seo/tags/16.8/vendor_prefixed?order=name

**Comment by miguelpeixe** [comment]
Hi @kbrown9!

Thank you for sharing this with us! I filed #539 so we can better discuss the issue.

**Pull request merged**